added recommendations with kafka

1EVILGUN1 opened the pull request:
добавление рекомендаций для пользователей. Использование kafka в рекомендациях. Измнение модуля stats, event,user
Сделано: схемы avro, protobuf, и collector sevice

1EVILGUN1 committed:
added recommendations with kafka

добавление рекомендаций для пользователей. Использование kafka в рекомендациях. Измнение модуля stats, event,user

1EVILGUN1 committed:
добавление аггрегатора

1EVILGUN1 committed:
добавление анализатора и изменение логики сбора статистики в event

Поле views у мероприятия заменили на rating с типом данных double.

1EVILGUN1 committed:
исправления в compilation

1EVILGUN1 committed:
remark to code part 1

изменение использования имен внутри сервисов, контроллеров. Подписывание методов и улучшение читаемость кода

1EVILGUN1 committed:
remark to code part 2

добавление настроек в конфиги к продюсерам и консьюмерам

1EVILGUN1 committed:
добавление логгирование в сервисы и контроллеры

1EVILGUN1 committed:
reformat code

1EVILGUN1 committed:
исправление checkstyle

VadimZharkov commented on core/event-service/src/main/java/ewm/compilation/model/Compilation.java:
`@Data `для сущностей не рекомендуется использовать
https://habr.com/ru/companies/haulmont/articles/564682/

1EVILGUN1 replied:
Тоже думал об этой идее, только было другое мнение по использованию Data для простых конструкций. При в дальнейшем при проектировании проектов к сведению

VadimZharkov commented on core/event-service/src/main/java/ewm/event/controller/pub/PublicEventController.java:
```diff
@@ -20,16 +18,40 @@
     }
 
     @GetMapping("/{id}")
     public UpdatedEventDto publicGetEvent(@PathVariable Long id, HttpServletRequest request) {
```
При обработке запроса к эндпоинту GET /events/{id} необходимо отправить информацию о просмотре пользователем мероприятия с идентификатором id. Идентификатор пользователя нужно взять из HTTP-заголовка X-EWM-USER-ID: `@RequestHeader("X-EWM-USER-ID") long userId`

VadimZharkov commented on core/event-service/src/main/java/ewm/event/controller/pub/PublicEventController.java:
```diff
@@ -20,16 +18,40 @@
-}
+
+    @GetMapping("/recommendations")
+    public List<RecommendationDto> getRecommendations(@RequestParam(defaultValue = "10") Long limit, HttpServletRequest request) {
```
Идентификатор пользователя передается в HTTP-заголовке X-EWM-USER-ID
Лучше в контроллере взять id из заголовока, и в сервис передать id

VadimZharkov commented on core/event-service/src/main/java/ewm/event/controller/pub/PublicEventController.java:
```diff
@@ -20,16 +18,40 @@
-        log.info("Получить события, согласно устловиям -> {}", requestParams);
-        return eventService.publicGetEvents(requestParams, request);
+        log.info("Получить события, согласно условиям -> {}", requestParams);
+        List<UpdatedEventDto> result = service.publicGetEvents(requestParams, request);
```
Тащить в класс-сервиса HttpServletRequest request не очень хорошая идея. Класс-сервиса работает с бизнес-сущностями и возможно с dto. HttpServletRequest request ни то, ни другое. Он больше относится с тех. деталям http протокола. Лучше из него на уровне контроллера взять нужную информацию и в сервис передать уже более осмысленные параметры.

VadimZharkov commented on stats/analyzer-service/src/main/java/ru/practicum/grpc/RecommendationsController.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public void getSimilarEvents(SimilarEventsRequestProto request, StreamObserver<RecommendedEventProto> responseObserver) {
+        List<RecommendedEventProto> response = service.getSimilarEvents(request);
```
Лучше добавить обработку ошибок
```
       try {
            List<RecommendedEventProto> events = service.getSimilarEvents(request);
            events.forEach(responseObserver::onNext);
            responseObserver.onCompleted();
        } catch (Exception e) {
            log.warn("Ошибка обработки запроса: {}", request, e);
            responseObserver.onError(e);
        }
```
Аналогично для остальных методов

VadimZharkov commented on stats/analyzer-service/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,17 @@
+create table if not exists event_similarity
+(
+    id         int8 GENERATED BY DEFAULT AS IDENTITY ( INCREMENT BY 1 MINVALUE 1 MAXVALUE 9223372036854775807 START 1 CACHE 1 NO CYCLE) NOT NULL,
+    event_a    int8,
```
NOT NULL

1EVILGUN1 replied:
переделал все поля на NOT NULL

VadimZharkov commented on stats/analyzer-service/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,17 @@
+(
+    id         int8 GENERATED BY DEFAULT AS IDENTITY ( INCREMENT BY 1 MINVALUE 1 MAXVALUE 9223372036854775807 START 1 CACHE 1 NO CYCLE) NOT NULL,
+    event_a    int8,
+    event_b    int8,
```
NOT NULL

VadimZharkov commented on stats/analyzer-service/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,17 @@
+    id         int8 GENERATED BY DEFAULT AS IDENTITY ( INCREMENT BY 1 MINVALUE 1 MAXVALUE 9223372036854775807 START 1 CACHE 1 NO CYCLE) NOT NULL,
+    event_a    int8,
+    event_b    int8,
+    score      float8,
```
NOT NULL

VadimZharkov commented on stats/analyzer-service/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,17 @@
+    event_a    int8,
+    event_b    int8,
+    score      float8,
+    event_time timestamp
```
NOT NULL

VadimZharkov reviewed (changes requested):
Хорошая работа, можно улучшить

1EVILGUN1 committed:
исправление замечаний ревьюера

1EVILGUN1 committed:
ошибки при компеляции проекта

1EVILGUN1 committed:
обновление публичного контроллера

1EVILGUN1 committed:
обновление публичного контроллера

добавление required = false к USER_ID_HEADER так как некоторые запросы идут без заголовка

1EVILGUN1 committed:
еще немного исправлений

1EVILGUN1 committed:
проправки
